### PR TITLE
Log home files reorganization

### PR DESCRIPTION
**Pull request by jshum2479**
This is for OWLS-97329 to provide option to organize the files in logHome differently.  By default, in this PR, files are organized by server name similar to on premise structures (servers/server name/logs/),  A logHomeLayout provides a way to set it back to current ways where all files are at the top level.  

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/10225/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/10281/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/10282/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/10283/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/10292/

### CODE REVIEW

**Comment by lennyphan** [comment]
Are ItIstioCrossClusters and ItIstioCrossClustersSetup are added as new files although they exist already?  Is this right?  Can't see its relationship to log home layout.

File: operator/src/main/resources/scripts/introspectDomain.py [hunk 1]
```diff
@@ -1804,7 +1808,14 @@ def customizeLog(self, name, bean, isDomainBean):
...
     self.writeln("<d:log" + logaction + ">")
     self.indent()
-    self.writeln("<d:file-name" + fileaction + ">" + logs_dir + "/" + name + ".log</d:file-name>")
+    if self.env.getDomainLogHomeLayout() == 'FLAT':
+      self.writeln("<d:file-name" + fileaction + ">" + logs_dir + "/" + name + ".log</d:file-name>")
+    else:
+      if not isDomainBean:
+        self.writeln("<d:file-name%s>%s/servers/%s/logs/%s.log</d:file-name>" % (fileaction, logs_dir, name, name))
+      else:
+        self.writeln("<d:file-name%s>%s/%s.log</d:file-name>" % (fileaction, logs_dir, name))
```
[the remark this answers is not in the document]
Comment (jshum2479) [reply operator/src/main/resources/scripts/introspectDomain.py:1817]: Yes,  the default behavior changed between 3.x and 4.0.    If user wants to use the old behavior, then they can set it to FLAT.   I am ok either way to go.

File: operator/src/main/resources/scripts/model_wdt_mii_filter.py [hunk 2]
```diff
@@ -278,7 +282,13 @@ def customizeLog(name, topologyOrServer):
   if 'Log' not in topologyOrServer:
     topologyOrServer['Log'] = {}
 
-  topologyOrServer['Log']['FileName'] = logs_dir + "/" + name + ".log"
+  if env.getDomainLogHomeLayout() == "FLAT":
```
[the remark this answers is not in the document]
Comment (jshum2479) [reply operator/src/main/resources/scripts/model_wdt_mii_filter.py:285]: Earlier in the code take cares of that.

 logs_dir = env.getDomainLogHome()
  if logs_dir is None or len(logs_dir) == 0:
    return

[the remark this answers is not in the document]
Comment (tbarnes-us) [reply operator/src/main/resources/scripts/model_wdt_mii_filter.py:285]: Cool - thanks

File: operator/src/main/resources/scripts/introspectDomain.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbarnes-us) [reply operator/src/main/resources/scripts/introspectDomain.py:1817]: I'm OK either way - but slightly lean towards 'old behavior' just to reduce the chance it surprises an existing user...

File: documentation/staging/content/userguide/managing-domains/_index.md [hunk 3]
```diff
@@ -166,6 +166,8 @@ The operator can automatically override WebLogic Server, domain, and introspecto
 This occurs if the Domain `logHomeEnabled` field is explicitly set to `true`, or if `logHomeEnabled` isn't set
 and `domainHomeSourceType` is set to `PersistentVolume`.  When overriding, the log location will be the location specified by the `logHome` setting.
 
+By default, all the log files are located follow the server layout subdirectories when of `logHome` and `logHomeEnabled` is setup.  If you want all the files to locate at a single level at the root of the `logHome`, you can specify `logHomeLayout` as `flat`.
```
[the remark this answers is not in the document]
Comment (jshum2479) [reply documentation/staging/content/userguide/managing-domains/_index.md]: updated

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java [hunk 4]
```diff
@@ -105,6 +106,19 @@ public class DomainSpec extends BaseConfiguration {
...
+   * The log files layout under logHome.
+   *   FLAT - all files is in one directory
+   *   BY_SERVERS - log files are organized under loghome/servers/server name/logs.
+   * */
+  @Description(
+      "Control how the log files under logHome is organized. "
+        + "FLAT - all files are under the logHome root directory. "
+        + "BY_SERVERS (default) - domain log file and introspector.out are at the logHome root level, all other files "
+        + "are organized under the respective server name logs directory.  logHome/servers/<server name>/logs.")
+  private LogHomeLayoutType logHomeLayout = LogHomeLayoutType.BY_SERVERS;
```
[the remark this answers is not in the document]
Comment (jshum2479) [reply operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java:121]: updated

File: documentation/staging/content/userguide/managing-domains/_index.md [hunk 5]
```diff
@@ -166,7 +166,33 @@ The operator can automatically override WebLogic Server, domain, and introspecto
 This occurs if the Domain `logHomeEnabled` field is explicitly set to `true`, or if `logHomeEnabled` isn't set
 and `domainHomeSourceType` is set to `PersistentVolume`.  When overriding, the log location will be the location specified by the `logHome` setting.
 
-By default, all the log files are located follow the server layout subdirectories when of `logHome` and `logHomeEnabled` is setup.  If you want all the files to locate at a single level at the root of the `logHome`, you can specify `logHomeLayout` as `flat`.
+Server log files are placed in a subdirectory `servers/<server name>/logs`
+by default when both `logHome` and `logHomeEnabled` are set.
+Set `logHomeLayout` to `FLAT` to place all log files at the `logHome` root.
+
+For example, the default layout of the log files under the `logHome` root.
```
Comment (tbarnes-us) [inline documentation/staging/content/userguide/managing-domains/_index.md:173]: For example, here is the default layout of the log files under the `logHome` root:

File: documentation/staging/content/userguide/managing-domains/domain-resource.md [hunk 6]
```diff
@@ -124,6 +124,7 @@ Elements related to logging:
 * `includeServerOutInPodLog`: Specifies whether the server .out file will be included in the Pod's log. Defaults to true.
 * `logHome`: The directory in a server's container in which to store the domain, Node Manager, server logs, server *.out, introspector .out, and optionally HTTP access log files if `httpAccessLogInLogHome` is true. Ignored if `logHomeEnabled` is false.
 * `logHomeEnabled`: Specifies whether the log home folder is enabled. Defaults to true if `domainHomeSourceType` is PersistentVolume; false, otherwise.
+* `logHomeLayout`: Specifies how the log files are organized under the `logHome` directory. Default to `BY_SERVER` or not specified, domain and introspector.out are placed at root level, all other log files are organized under logHome/servers/server name/logs. If `FLAT` is specified, all files are placed at root level.
```
Comment (tbarnes-us) [inline documentation/staging/content/userguide/managing-domains/domain-resource.md:127]: Needs back-ticks, and slight fixes:

```
Specifies how log files are organized under the `logHome` directory when `logHomeEnabled` is set to `true`.
Defaults to `BY_SERVER`, where domain logs and `introspector.out` are placed at root level
and all other log files are organized under `logHome/servers/<server-name>/logs`.
If `FLAT` is specified, then all files are placed at root level.
```

File: documentation/staging/content/userguide/managing-domains/domain-resource.md [hunk 7]
```diff
@@ -124,6 +124,7 @@ Elements related to logging:
 * `includeServerOutInPodLog`: Specifies whether the server .out file will be included in the Pod's log. Defaults to true.
 * `logHome`: The directory in a server's container in which to store the domain, Node Manager, server logs, server *.out, introspector .out, and optionally HTTP access log files if `httpAccessLogInLogHome` is true. Ignored if `logHomeEnabled` is false.
```
Comment (tbarnes-us) [inline documentation/staging/content/userguide/managing-domains/domain-resource.md:125]: Append 
```
See also `logHomeLayout`.
```

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java [hunk 8]
```diff
@@ -112,10 +113,11 @@ public class DomainSpec extends BaseConfiguration {
    *   BY_SERVERS - log files are organized under loghome/servers/server name/logs.
    * */
   @Description(
-      "Control how the log files under logHome is organized. "
-        + "FLAT - all files are under the logHome root directory. "
-        + "BY_SERVERS (default) - domain log file and introspector.out are at the logHome root level, all other files "
-        + "are organized under the respective server name logs directory.  logHome/servers/<server name>/logs.")
+      "Control how log files under `logHome` are organized when logHome is set and `logHomeEnabled` is true. "
+        + "`FLAT` - - all files directly in the `logHome` root directory. "
```
Comment (tbarnes-us) [inline operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java:117]: Second logHome is missing back-ticks.

Maybe change "- -" to "-".

directory. --> directory (remove period)

logHome/servers/<server name>/logs is missing back-ticks

**Review by rjeberhard — APPROVED** [review]
Approving, assuming agreement on necessary changes to the field help text.

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java (hunk 8 above)
Comment (jshum2479) [reply operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java]: fixed

File: documentation/staging/content/userguide/managing-domains/domain-resource.md (hunk 7 above)
Comment (jshum2479) [reply documentation/staging/content/userguide/managing-domains/domain-resource.md]: fixed

File: documentation/staging/content/userguide/managing-domains/domain-resource.md (hunk 6 above)
Comment (jshum2479) [reply documentation/staging/content/userguide/managing-domains/domain-resource.md]: fixed

**Review by lennyphan — APPROVED** [review]
LGTM

File: operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java [hunk 9]
```diff
@@ -102,9 +103,24 @@ public class DomainSpec extends BaseConfiguration {
...
+   * The log files layout under logHome.
+   *   FLAT - all files is in one directory
+   *   BY_SERVERS - log files are organized under loghome/servers/server name/logs.
+   * */
+  @Description(
+      "Control how log files under `logHome` are organized when logHome is set and `logHomeEnabled` is true. "
+        + "`FLAT` - all files directly in the `logHome` root directory. "
+        + "`BY_SERVERS` (default) - domain log files and `introspector.out` are at the `logHome` root level, "
+        + "all other files are organized under the respective server name logs directory  "
+        + "logHome/servers/<server name>/logs.")
```
Comment (tbarnes-us) [inline operator/src/main/java/oracle/kubernetes/weblogic/domain/model/DomainSpec.java:120]: minor nit: logHome/servers/<server name>/logs should be encased in back-ticks

**Review by tbarnes-us — APPROVED** [review]
Approved, although still missing some backticks.

**Pull request merged**